sound/soc/sof/sof-audio.c

```diff
@@ -307,23 +307,37 @@ int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, in
 	const struct ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg;
 	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
 	struct snd_soc_dapm_widget *widget;
-	int i, ret, num_widgets;
+	int i, ret, num_widgets = 0;
+	int start, step;
 
 	/* nothing to set up */
 	if (!list)
 		return 0;
 
+	/* set up DAI widget first for capture */
+	if (dir == SOF_IPC_STREAM_CAPTURE) {
+		start = list->num_widgets - 1;
+		step = -1;
+	} else {
+		start = 0;
+		step = 1;
+	}
+
 	/* set up widgets in the list */
-	for_each_dapm_widgets(list, num_widgets, widget) {
-		struct snd_sof_widget *swidget = widget->dobj.private;
+	for (i = start; i < list->num_widgets && i >= 0; i += step) {
+		struct snd_sof_widget *swidget;
 
+		widget = list->widgets[i];
+		swidget = widget->dobj.private;
 		if (!swidget)
 			continue;
 
 		/* set up the widget */
 		ret = sof_widget_setup(sdev, swidget);
 		if (ret < 0)
 			goto widget_free;
+
+		num_widgets++;
```

Comment:
it looks like this patch changes the handling of num_widgets.

In the previous version with for_each_dapm_widgets(), num_widgets was incremented at every loop iteration, but decremented in the error handler conditionally. That wasn't balanced.

It's no longer the case, so wondering if we should first have a fix for the loop handling, then support a change of direction?

Reply:
@plbossart num_widgets was incremented in for_each)dapm_widgets() just like it is here. It was definitely balanced.

Reply:
no, I maintain it was different. The macro incremented num_widgets for every iteration, while here it's incremented if the continue and goto branches are not taken.